### PR TITLE
15211: search pronouns

### PR DESCRIPTION
**Pull request by gedu**
### Details

Now the `PronounsPage` is a class component and will show a search input to filter the pronouns.

### Fixed Issues

$ https://github.com/Expensify/App/issues/15211
PROPOSAL: https://github.com/Expensify/App/issues/15211#issuecomment-1436578720

### Tests

1. Open the Pronouns page
2. You should see an input and no options

1. Open the Pronouns page
2. Write some letters to start seeing some options
3. Click on one and it should be applied to your profile

- [ ] Verify that no errors appear in the JS console

### Offline tests

1. Open the Pronouns page
2. Write some letters to start seeing some options
3. Click on one and it should be applied to your profile only locally, you won't see it applied to other devices
4. Connect to the internet, it will sync with the server and you will see the selected pronouns on your other devices

### QA Steps

Same as above

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/1676818/220338744-bd3fea6b-6be6-4478-b1c4-5347c46a264d.mp4

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/1676818/220338815-3e6118cd-51d8-40d0-a95e-7992e2cc463a.mp4

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/1676818/220338920-aa918ea4-bd10-4dbd-b3aa-bb5e6386ae92.mp4

</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/1676818/220338988-848230b9-1487-4a02-89f3-b026e0e2ccf1.mp4

</details>

<details>
<summary>iOS</summary>

https://user-images.githubusercontent.com/1676818/220339195-6fd72d7d-1ba4-4535-9052-619913704046.mp4

</details>

<details>
<summary>Android</summary>

https://user-images.githubusercontent.com/1676818/220339312-44df8e69-4720-4da6-8e8b-3c84a189a5f8.mp4

</details>

### CODE REVIEW

**Comment by puneetlath** [comment]
@gedu are you ready for us to review? If so, I'll let @mollfpr take the first pass.

**Review by mollfpr — COMMENTED** [review]
Missing JSDoc on _onChangeText_ function.

**Comment by gedu** [comment]
> @gedu are you ready for us to review? If so, I'll let @mollfpr take the first pass.

Yes, should be ready

**Comment by gedu** [comment]
> Missing JSDoc on _onChangeText_ function.

What should I add? I follow [Avoid descriptions that don't add any additional information](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)
it doesn't receive a param and doesn't return a value.  I looked at `SearchPage`, `NewChatPage`, `WorkspaceInvitePage` and didn't have JSDoc. 

Could you help me what would be the best way to Document a function like `onChangeText`?

**Comment by mollfpr** [comment]
Yeah, you’re right! Since we have a default value for the parameter, we’re good without the JSDoc.

**Comment by mollfpr** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android / native
    - [ ] Android / Chrome
    - [ ] iOS / native
    - [ ] iOS / Safari
    - [ ] MacOS / Chrome / Safari
    - [ ] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>



</details>

<details>
<summary>Mobile Web - Chrome</summary>



</details>

<details>
<summary>Mobile Web - Safari</summary>



</details>

<details>
<summary>Desktop</summary>



</details>

<details>
<summary>iOS</summary>



</details>

<details>
<summary>Android</summary>



</details>

**Comment by puneetlath** [comment]
Good question. Maybe we should do it similar to the Timezone page. Where the input is filled with your current choice, but the text is selected so that if you start typing it is overwritten. What do you think? 

https://user-images.githubusercontent.com/1371996/220680871-c72ead61-cf16-4c04-bf29-ab28c88b1bdb.mov

**Comment by puneetlath** [comment]
Also, we should allow selection via arrow keys. Like we do on the New Chat page.

**Comment by gedu** [comment]
> Good question. Maybe we should do it similar to the Timezone page. Where the input is filled with your current choice, but the text is selected so that if you start typing it is overwritten. What do you think?

I Think this is easy to add. 

> Also, we should allow selection via arrow keys. Like we do on the New Chat page.

For this one I will have to look a bit more

**Comment by gedu** [comment]
@puneetlath for the input filled seems that there is a bug on iOS, it isn't selecting the text (this error happens on Timezone Page too)


https://user-images.githubusercontent.com/1676818/220884035-6dd05737-741b-4a4b-a95c-45c54c0be626.mov

what should I do?

**Comment by mollfpr** [comment]
@gedu Is the issue you mentioned need to be fixed separately on TimezonePage and PronounsPage? Or it can be in one fix?

**Comment by gedu** [comment]
> @gedu Is the issue you mentioned need to be fixed separately on TimezonePage and PronounsPage? Or it can be in one fix?

It can be fixed together, I rollback and won't submit it

**Comment by gedu** [comment]
All should be ok now, is behaving the same way as the videos in the description

**Comment by puneetlath** [comment]
ok cool. @mollfpr can you give it the first review?

**Comment by mollfpr** [comment]
The code changes look good.


@puneetlath Do you think the bottom margin of the title and the selection is too far?

| Before | After |
|------------|-------|
|Screen Shot 2023-02-25 at 01 29 06|Screen Shot 2023-02-25 at 01 28 58|

**Comment by mollfpr** [comment]
We have extra 12px vertical padding on the text input; we can add new props for the text input wrapper to eliminate the top padding.

Here's how it looks like before/after adding 0 top padding on the text input wrapper.
| Before | After |
|------------|-------|
|Screen Shot 2023-02-25 at 01 28 58|Screen Shot 2023-02-25 at 01 53 19|

**Review by puneetlath — DISMISSED** [review]

**Comment by shawnborton** [comment]
I like the after shot where you remove some of that padding.

Also, can we make the search results look like the list selection used to look like, with the horizontal separators between results?
image


Also, I agree with having the current pronoun look exactly like the timezone does. Ideally that should be a reusable push-to-page component we can use.

**Comment by puneetlath** [comment]
Ok cool. So to summarize @gedu we want to: 

1. Update the padding
2. Style it exactly like the timezone page, with separators

**Comment by puneetlath** [comment]
Looks like that just means we need to pass the `shouldHaveOptionSeparator` prop on the `OptionsSelector`.

**Comment by gedu** [comment]
> Ok cool. So to summarize @gedu we want to:
> 
> 1. Update the padding
> 2. Style it exactly like the timezone page, with separators

Cool, I will take a look at it

**Comment by gedu** [comment]
@mollfpr Hey, I saw that the text `Your pronouns are shown on your profile.` had an extra 6px vertical margin, removing it the screen looks like this

Screen Shot 2023-02-27 at 13 21 04

What do you think? or should I add a new prop to override the style for the input? 

Note: if I add a new prop I think is better to pass a style and let you override, because playing with just setting the padding to 0px could make the component confusing, because you can place the input at the bottom and the vertical padding, in that case, is different @puneetlath

**Comment by shawnborton** [comment]
Hmm I'm not sure what the standard is here or if we are being consistent throughout the app, but I think 12px of top padding in these views (where you have a page header followed by text or a UI component) typically looks nice:
image

I think we removed the top padding on views that start with an avatar though, so maybe we don't have a consistent pattern here yet.

**Comment by mollfpr** [comment]
@gedu @shawnborton Yeah, we don't have a consistent pattern yet. I'm okay if we go with the current looks until we have the correct margin between the title and the component.

**Comment by gedu** [comment]
> @gedu @shawnborton Yeah, we don't have a consistent pattern yet. I'm okay if we go with the current looks until we have the correct margin between the title and the component.

So, should I remove the extra margin from the title? `styles.mb6`

**Comment by shawnborton** [comment]
For that change, can we see what it looks like when the input focused? We probably want 20px of margin under the text and above the input.

**Comment by mollfpr** [comment]
@shawnborton Here you go.

Screen Shot 2023-02-27 at 23 28 56

**Comment by shawnborton** [comment]
Cool, that works for me.

**Comment by mollfpr** [comment]
@gedu I found a bug after refreshing the pronouns page; then selecting a pronoun. The selected pronoun is not checked, and the input value is not cleared.

https://user-images.githubusercontent.com/25520267/221626386-6b6aa5af-b02f-48d0-8962-214c6bd430b2.mov

**Comment by gedu** [comment]
> @gedu I found a bug after refreshing the pronouns page; then selecting a pronoun. The selected pronoun is not checked, and the input value is not cleared.

I will take a look

**Comment by gedu** [comment]
> @gedu I found a bug after refreshing the pronouns page; then selecting a pronoun. The selected pronoun is not checked, and the input value is not cleared.

I'm still looking at this, this bug is happening on `TimezonePage` too. Seems that after refreshing the page, the lifecycle of the component breaks. 

@mountiny should I keep looking into this? or we can create a follow-up PR given that happens on other pages?

**Comment by mollfpr** [comment]
@gedu On the timezone select page, it's fixed on this [PR](https://github.com/Expensify/App/pull/15176). Maybe you can find something there?

**Comment by mountiny** [comment]
@gedu Lets have a look at the other PR @mollfpr linked if we can see how it was fixed there. Would be ideal to do it here too.

**Comment by gedu** [comment]
> @gedu Lets have a look at the other PR @mollfpr linked if we can see how it was fixed there. Would be ideal to do it here too.

Yes, I just found a similar solution and is working, thanks for the input, will upload the fix soon

**Comment by gedu** [comment]
https://user-images.githubusercontent.com/1676818/221840672-b4d1d851-a690-462a-b1d2-371d12117e03.mp4

@mollfpr

File: src/pages/settings/Profile/PronounsPage.js [hunk 1]
```diff
@@ -27,60 +27,117 @@ const defaultProps = {
...
+                customIcon: isCurrentPronouns ? greenCheckmark : undefined,
+
+                // This property will make the currently selected value have bold text
+                boldStyle: isCurrentPronouns,
+            };
+        });
+    }
+
+    /**
+     * @param {Obj} selectedPronouns
```
Comment (mountiny) [inline src/pages/settings/Profile/PronounsPage.js:99]:
```suggestion
     * @param {Object} selectedPronouns
```

**Review by mountiny — COMMENTED** [review]
@gedu couple of comments

@mollfpr will you be able to run through the checklist today 💪 ?

File: src/pages/settings/Profile/PronounsPage.js [hunk 2]
```diff
@@ -30,33 +30,26 @@ const defaultProps = {
...
+        this.loadPronouns();
         this.state = {
             searchValue: '',
         };
     }
 
+    componentDidUpdate(prevProps) {
+        if (prevProps.currentUserPersonalDetails.pronouns === this.props.currentUserPersonalDetails.pronouns) {
+            return;
+        }
```
[the remark this answers is not in the document]
Comment (gedu) [reply src/pages/settings/Profile/PronounsPage.js:47]: done

File: src/pages/settings/Profile/PronounsPage.js (hunk 1 above)
Comment (gedu) [reply src/pages/settings/Profile/PronounsPage.js:99]: done

**Comment by mollfpr** [comment]
Thanks, @gedu for the updates! @mountiny Yup, I’ll review it again within an hour.

**Review by puneetlath — APPROVED** [review]
Code looks good to me.

**Review by mollfpr — APPROVED** [review]
LGTM, good work @gedu 👍

**Pull request merged**